File: src/lib/tags.ts
/**
 * tag 校验与纯 JSON 变换（可被 Client Component 安全导入）。
 * 写库（renameAcrossRecords）在 `@/lib/tagsdb`，勿在此文件 import `@/db`。
 */
import { newInternalMsg } from '@/lib/myerr' 

/**
 * 验证 tag 格式
 * 规则：
 * - 只能用：英文字母、数字、下划线、冒号
 * - 不能以数字开头
 * - 冒号不能开头、不能结尾、不能连续
 * - 至少一个字符
 * - 类似标识符命名规则，允许冒号表示层级（如 source:device、review:weekly）
 */
export function isValidTag(tag: string): boolean {
  return /^[a-zA-Z_][a-zA-Z0-9_]*(?::[a-zA-Z0-9_]+)*$/.test(tag)
}

/**
 * 保留 tag **前缀**列表（非仅精确匹配）。
 * 某 tag 视为保留当且仅当：`tag === P` 或 `tag.startsWith(P + ":")`
 *（冒号边界，避免误伤 `transaction_entrypoint`）。
 * 当前 P：`transaction_entry`、`body:weight`、`todo`、`review`。
 * 仅专用 API 可写入带此前缀的 tag；通用 log / Admin 草稿 / rename 的 from/to 均拒绝。
 */
export const RESERVED_TAG_PREFIXES = [
  'transaction_entry',
  'body:weight',
  'todo',
  'review',
] as const

export type ReservedTagPrefix = (typeof RESERVED_TAG_PREFIXES)[number]

export const RESERVED_TAG_TRANSACTION_ENTRY: ReservedTagPrefix = 'transaction_entry'
export const RESERVED_TAG_BODY_WEIGHT: ReservedTagPrefix = 'body:weight'
export const RESERVED_TAG_TODO: ReservedTagPrefix = 'todo'
export const RESERVED_TAG_REVIEW: ReservedTagPrefix = 'review'

/** 保留 tag 错误后缀：不指向具体端点路径，AI 自行查 OpenAPI（端点改名/新增不会过时） */
const RESERVED_TAG_HINT =
  'use the dedicated log API for this record type'

/** 组装落库用的类型 tag：`transaction_entry:income` / `transaction_entry:expense` */
export function transactionEntryTypeTag(
  type: 'income' | 'expense',
): string {
  return `${RESERVED_TAG_TRANSACTION_ENTRY}:${type}`
}

export function isReservedTag(tag: string): boolean {
  for (const p of RESERVED_TAG_PREFIXES) {
    if (tag === p || tag.startsWith(`${p}:`)) {
      return true
    }
  }
  return false
}

/** 英文错误：指明保留 tag 应走专用记录 API（不指向具体端点） */
export function reservedTagError(tag: string): string {
  return `tag "${tag}" is reserved; ${RESERVED_TAG_HINT}`
}

/** 与 Go `tags.ValidationResult` 同构 */
export type ValidationResult = { valid: boolean; error?: string }

/**
 * 客户端传入的 tags 不得含保留前缀。
 * 服务端写入的 transaction 行可含保留 tag，不要对本函数传入那些组装结果来「拒绝」。
 */
export function assertNoReservedTags(tags: string[]): ValidationResult {
  for (const tag of tags) {
    if (isReservedTag(tag)) {
      return { valid: false, error: reservedTagError(tag) }
    }
  }
  return { valid: true }
}

/**
 * 验证 tags 数组
 * - 可为空数组（[] 合法）
 * - 每个 tag 都符合格式
 */
export function validateTags(tags: string[]): ValidationResult {
  if (!Array.isArray(tags)) {
    return { valid: false, error: 'tags must be an array of strings' }
  }

  for (const tag of tags) {
    if (!isValidTag(tag)) {
      return {
        valid: false,
        error: `invalid tag: "${tag}". Tags must contain only letters, numbers, underscores, and cannot start with a number.`,
      }
    }
  }

  return { valid: true }
}

/**
 * 返回 tags 中第一个重复的 tag 名；无重复返回 null。
 * 各写入端点（numbers/text/todo/body/weight/review）在落库前调用，重复 → 400。
 * 文案由调用方拼：`duplicate tag "${tag}"`（batch 端点加 `entries[i]:` 前缀）。
 */
export function firstDuplicateTag(tagList: string[]): string | null {
  const seen = new Set<string>()
  for (const tag of tagList) {
    if (seen.has(tag)) {
      return tag
    }
    seen.add(tag)
  }
  return null
}

/** 与 Go `tags.ErrTagsNotJSONArray` 同文案：解析成功但根不是数组 */
export const TAGS_NOT_JSON_ARRAY = 'tags field is not a JSON array'

/** 解析 records.tags JSON；非法 JSON / 根非数组 = DB 脏数据 → 空数组
 * （聚合时静默跳过该行，与 rename 的 FromDB 兜底语义统一——2026-08-06 用户拍板）。
 * 与 Go `tags.parseTagsJSONArray` 对称。 */
function parseTagsJsonArray(tagsJson: string): unknown[] {
  try {
    const parsed: unknown = JSON.parse(tagsJson)
    return Array.isArray(parsed) ? parsed : []
  } catch {
    return []
  }
}

/** 单个 tag 的计数（JSON `tag`/`count` snake_case） */
export type TagCount = { tag: string; count: number }

/**
 * 从 records.tags（JSON 字符串数组）汇总「记录条数」，按「计数降序、同名 tag 升序」返回数组。
 * prefix 非空时仅保留以 prefix 开头的 tag（真前缀，自动补全语义）。
 * 排序与 Go `sort.Slice`（计数降序，同名升序）一致；tag 名比较用字节序（ASCII 下大写在小写前），勿用 localeCompare。
 * 非法 JSON / 非数组抛错（由 HTTP 映射 500）。
 */
export function aggregateTagCounts(
  tagFields: string[],
  prefix = '',
): TagCount[] {
  const counts = new Map<string, number>()

  for (const field of tagFields) {
    const parsed = parseTagsJsonArray(field)
    for (const tag of parsed) {
      if (typeof tag !== 'string') continue
      counts.set(tag, (counts.get(tag) ?? 0) + 1)
    }
  }

  const list: TagCount[] = []
  for (const [tag, count] of counts) {
    if (prefix && !tag.startsWith(prefix)) continue
    list.push({ tag, count })
  }
  list.sort((a, b) => {
    if (a.count !== b.count) return b.count - a.count
    return a.tag < b.tag ? -1 : a.tag > b.tag ? 1 : 0
  })
  return list
}

/**
 * 数组版 rename 变换（与 Go `tags.renameTags` 对称，§10b 步骤 2 二次定案）：
 * to ∈ tags → 移除 from（去重语义）；否则 from 原位替换为 to；from ∉ tags → 返回 null（防御）。
 */
export function renameTags(
  tags: string[],
  from: string,
  to: string,
): string[] | null {
  const fromIdx = tags.indexOf(from)
  if (fromIdx < 0) return null
  if (tags.includes(to)) {
    return tags.filter((t) => t !== from)
  }
  return tags.map((t) => (t === from ? to : t))
}

/**
 * rename 业务校验：非空、合法 tag、非保留、from≠to。
 * 与 Go `ValidateRename` 同构；调用方应先 trim。
 */
export function validateRename(from: string, to: string): ValidationResult {
  if (!from || !to) {
    return { valid: false, error: 'missing required fields: from, to' }
  }
  if (!isValidTag(from) || !isValidTag(to)) {
    return { valid: false, error: 'from and to must be valid tag names' }
  }
  if (isReservedTag(from)) {
    return { valid: false, error: reservedTagError(from) }
  }
  if (isReservedTag(to)) {
    return { valid: false, error: reservedTagError(to) }
  }
  if (from === to) {
    return { valid: false, error: 'from and to must be different' }
  }
  return { valid: true }
}

File: faas/internal/tags/tags.go
package tags

import (
	"context"
	"fmt"
	"regexp"
	"sort"
	"strings"

	"github.com/mdk/digitaltwin2026/faas/internal/db"
	"github.com/mdk/digitaltwin2026/faas/internal/myerr"
	"github.com/mdk/digitaltwin2026/faas/internal/recordrepo"
)

var tagPattern = regexp.MustCompile(`^[a-zA-Z_][a-zA-Z0-9_]*(?::[a-zA-Z0-9_]+)*$`)

// ReservedTagPrefixes 保留 tag **前缀**列表（非仅精确匹配）。
// 某 tag 视为保留当且仅当：tag == P 或 strings.HasPrefix(tag, P+":")
// （冒号边界，避免误伤 transaction_entrypoint）。
// 当前 P：transaction_entry、body:weight、todo、review。
// 仅专用 API 可写入带此前缀的 tag；通用 log / Admin 草稿 / rename 的 from/to 均拒绝。
var ReservedTagPrefixes = []string{"transaction_entry", "body:weight", "todo", "review"}

const ReservedTagTransactionEntry = "transaction_entry"
const ReservedTagBodyWeight = "body:weight"
const ReservedTagTodo = "todo"
const ReservedTagReview = "review"

// reservedTagHint 保留 tag 错误后缀：不指向具体端点路径，AI 自行查 OpenAPI
// （端点改名/新增不会过时；与 TS RESERVED_TAG_HINT 同句）。
const reservedTagHint = "use the dedicated log API for this record type"

// TransactionEntryTypeTag 组装落库用类型 tag。
func TransactionEntryTypeTag(typ string) string {
	return ReservedTagTransactionEntry + ":" + typ
}

func IsValidTag(tag string) bool {
	return tagPattern.MatchString(tag)
}

func IsReservedTag(tag string) bool {
	for _, p := range ReservedTagPrefixes {
		if tag == p || strings.HasPrefix(tag, p+":") {
			return true
		}
	}
	return false
}

// InvalidTagMessage 单个 tag 非法文案（与 ValidateTags 数组版同文案；tags-add handler 用）。
func InvalidTagMessage(tag string) string {
	return fmt.Sprintf(
		`invalid tag: "%s". Tags must contain only letters, numbers, underscores, and cannot start with a number`,
		tag,
	)
}

// ReservedTagError 英文错误：指明保留 tag 应走专用记录 API（不指向具体端点）。
func ReservedTagError(tag string) string {
	return fmt.Sprintf(`tag "%s" is reserved; %s`, tag, reservedTagHint)
}

type ValidationResult struct {
	Valid bool
	Error string
}

func AssertNoReservedTags(tagList []string) ValidationResult {
	for _, tag := range tagList {
		if IsReservedTag(tag) {
			return ValidationResult{Valid: false, Error: ReservedTagError(tag)}
		}
	}
	return ValidationResult{Valid: true}
}

func ValidateTags(tags []string) ValidationResult {
	for _, tag := range tags {
		if !IsValidTag(tag) {
			return ValidationResult{Valid: false, Error: InvalidTagMessage(tag)}
		}
	}
	return ValidationResult{Valid: true}
}

// FirstDuplicateTag 返回 tags 中第一个重复的 tag 名；无重复返回 ""。
// 各写入端点（numbers/text/todo/body/weight/review）在落库前调用，重复 → 400。
// 文案由调用方拼：Duplicate tag "<tag>"（batch 端点加 entries[i]: 前缀）。
func FirstDuplicateTag(tagList []string) string {
	seen := make(map[string]struct{}, len(tagList))
	for _, tag := range tagList {
		if _, ok := seen[tag]; ok {
			return tag
		}
		seen[tag] = struct{}{}
	}
	return ""
}

// ValidateNormalize normalize 业务校验（零 DB，调用方应先 trim 元素）：
// from 非空数组（元素合法 / 无重复 / 非保留前缀）、to 非空合法非保留、to ∉ from。
// 校验顺序定案（docs/20260805-tag-design.md §tag 归一化）：from 形状 → to 缺失 →
// from 元素（逐个：非法 → 重复 → 保留）→ to（非法 → 保留）→ 交集。
func ValidateNormalize(from []string, to string) ValidationResult {
	if len(from) == 0 {
		return ValidationResult{Valid: false, Error: "missing required field: from"}
	}
	if to == "" {
		return ValidationResult{Valid: false, Error: "missing required field: to"}
	}
	seen := map[string]bool{}
	for _, f := range from {
		if !IsValidTag(f) {
			return ValidationResult{Valid: false, Error: InvalidTagMessage(f)}
		}
		if seen[f] {
			return ValidationResult{Valid: false, Error: fmt.Sprintf(`duplicate tag in from: "%s"`, f)}
		}
		seen[f] = true
		if IsReservedTag(f) {
			return ValidationResult{Valid: false, Error: ReservedTagError(f)}
		}
	}
	if !IsValidTag(to) {
		return ValidationResult{Valid: false, Error: InvalidTagMessage(to)}
	}
	if IsReservedTag(to) {
		return ValidationResult{Valid: false, Error: ReservedTagError(to)}
	}
	if seen[to] {
		return ValidationResult{Valid: false, Error: "to must not be in from"}
	}
	return ValidationResult{Valid: true}
}

// ErrTagsNotJSONArray 与 TS TAGS_NOT_JSON_ARRAY 同文案（transactions-summary 行解析仍 500 用）。
const ErrTagsNotJSONArray = "tags field is not a JSON array"

// TagCount 单个 tag 的计数（JSON `tag`/`count` snake_case）。
type TagCount struct {
	Tag   string `json:"tag"`
	Count int    `json:"count"`
}

// AggregateTagCounts 汇总 tag 出现次数，按「计数降序、同名 tag 升序」返回。
// prefix 非空时仅保留 strings.HasPrefix(tag, prefix) 的 tag（真前缀，自动补全语义）。
// 非法 JSON / 非数组返回 error（HTTP 映射 500）。
func AggregateTagCounts(tagLists [][]string, prefix string) []TagCount {
	counts := map[string]int{}
	for _, tags := range tagLists {
		for _, tag := range tags {
			counts[tag]++
		}
	}
	list := make([]TagCount, 0, len(counts))
	for tag, count := range counts {
		if prefix != "" && !strings.HasPrefix(tag, prefix) {
			continue
		}
		list = append(list, TagCount{Tag: tag, Count: count})
	}
	sort.Slice(list, func(i, j int) bool {
		if list[i].Count != list[j].Count {
			return list[i].Count > list[j].Count
		}
		return list[i].Tag < list[j].Tag
	})
	return list
}

// NormalizePageSize normalize 分页循环的页大小（写死 100，§6 定案；原 rename 同名常量）。
const NormalizePageSize = 100

// NormalizeAcrossRecords 在单事务内将 tags 中 from 系列归一化为 to（业务层编排，
// §10b 步骤 2 骨架复用 + normalize 定案）：
// uow 开事务 → repo.AcquireRenameLock（advisory xact lock：并发互斥、随事务结束自动释放）
// → 对每个 from 元素分页扫描（FindByCriteria 的 Tags 为 AND 交集语义，不能一次匹配任一；
// 每行 normalizeTags 做一次多源变换——顺序语义定案「删 from 系列 + 尾加 to」，
// 不能分解为逐源 rename）→ 命中才 repo.Update 写回 → len(页) < NormalizePageSize 终止。
// 同一行含多个 from 元素时：首次命中更新，后续 normalizeTags 返回 !changed 不重复写。
// 中途失败全滚（任何 DB 错误 → 500）；OFFSET 分页 + 事务内多页：页间并发提交可能跳行/漏改（尽力而为）。
func NormalizeAcrossRecords(ctx context.Context, b db.TxBeginner, from []string, to string) (int, *myerr.MyError) {
	updated := 0
	me := db.WithTx(ctx, b, func(q db.Executor) *myerr.MyError {
		if me := recordrepo.Repo.AcquireRenameLock(ctx, q); me != nil {
			return me
		}
		for _, f := range from {
			page := 1
			for {
				recs, me := recordrepo.Repo.FindByCriteria(ctx, q, recordrepo.FindCriteria{
					Criteria:  recordrepo.Criteria{Tags: []string{f}},
					Page:      page,
					PageSize:  NormalizePageSize,
					SortBy:    "id",
					SortOrder: "asc",
				})
				if me != nil {
					return me
				}
				for _, rec := range recs {
					next, ok := normalizeTags(rec.Tags, from, to)
					if !ok {
						continue
					}
					rec.Tags = next
					if me := recordrepo.Repo.Update(ctx, q, rec); me != nil {
						return me
					}
					updated++
				}
				if len(recs) < NormalizePageSize {
					break
				}
				page++
			}
		}
		return nil
	})
	if me != nil {
		return 0, me
	}
	return updated, nil
}

// normalizeTags 多源一次变换（normalize 定案；替代 renameTags 串版）：
// tags 含 from 中任意 tag → 全部原地删（后续前移）；to 已存在（且非 from 元素）保持原位，
// 否则尾加。from 含 to 已被 ValidateNormalize 拦截（防御：若发生，先删后尾加）。
// 不含任何 from 元素 → 原样返回 changed=false。
func normalizeTags(tags []string, from []string, to string) ([]string, bool) {
	inFrom := make(map[string]bool, len(from))
	for _, f := range from {
		inFrom[f] = true
	}
	changed := false
	for _, t := range tags {
		if inFrom[t] {
			changed = true
			break
		}
	}
	if !changed {
		return tags, false
	}
	out := make([]string, 0, len(tags))
	hasTo := false
	for _, t := range tags {
		if inFrom[t] {
			continue // 原地删、后续前移
		}
		out = append(out, t)
		if t == to {
			hasTo = true
		}
	}
	if !hasTo {
		out = append(out, to) // target 不存在 → 尾加
	}
	return out, true
}
